python/src/nanoarrow/_lib.pyx

```diff
@@ -478,6 +593,130 @@ cdef class CSchemaView:
     def __repr__(self):
         return _lib_utils.schema_view_repr(self)
 
+
+cdef class CSchemaBuilder:
+    cdef CSchema c_schema
+    cdef ArrowSchema* _ptr
+
+    def __cinit__(self, CSchema schema):
+        self.c_schema = schema
+        self._ptr = schema._ptr
+        if self._ptr.release == NULL:
+            ArrowSchemaInit(self._ptr)
+
+    @staticmethod
+    def allocate():
+        return CSchemaBuilder(CSchema.allocate())
+
+    def child(self, int64_t i):
+        return CSchemaBuilder(self.c_schema.child(i))
+
+    def set_type(self, int type):
+        self.c_schema._assert_valid()
+
+        cdef int result = ArrowSchemaSetType(self._ptr, <ArrowType>type)
+        if result != NANOARROW_OK:
+            Error.raise_error("ArrowSchemaSetType()", result)
+
+        return self
+
+    def set_type_decimal(self, int type, int precision, int scale):
+        self.c_schema._assert_valid()
+
+        cdef int result = ArrowSchemaSetTypeDecimal(self._ptr, <ArrowType>type, precision, scale)
+        if result != NANOARROW_OK:
+            Error.raise_error("ArrowSchemaSetType()", result)
+
+    def set_type_fixed_size(self, int type, int fixed_size):
+        self.c_schema._assert_valid()
+
+        cdef int result = ArrowSchemaSetTypeFixedSize(self._ptr, <ArrowType>type, fixed_size)
+        if result != NANOARROW_OK:
+            Error.raise_error("ArrowSchemaSetTypeFixedSize()", result)
+
+        return self
+
+    def set_type_date_time(self, int type, int time_unit, timezone):
+        self.c_schema._assert_valid()
+
+        cdef int result
+        if timezone is None:
+            result = ArrowSchemaSetTypeDateTime(self._ptr, <ArrowType>type, <ArrowTimeUnit>time_unit, NULL)
+        else:
+            timezone = str(timezone)
+            result = ArrowSchemaSetTypeDateTime(self._ptr, <ArrowType>type, <ArrowTimeUnit>time_unit, timezone.encode("UTF-8"))
+
+        if result != NANOARROW_OK:
+            Error.raise_error("ArrowSchemaSetTypeDateTime()", result)
+
+        return self
+
+    def set_format(self, str format):
+        self.c_schema._assert_valid()
+
+        cdef int result = ArrowSchemaSetFormat(self._ptr, format.encode("UTF-8"))
+        if result != NANOARROW_OK:
+            Error.raise_error("ArrowSchemaSetFormat()", result)
+
+        return self
+
+    def set_name(self, name):
+        self.c_schema._assert_valid()
+
+        cdef int result
+        if name is None:
+            result = ArrowSchemaSetName(self._ptr, NULL)
```

Comment:
FWIW, it seems that Arrow C++ sets an empty string by default for schemas (types) that don't have a name, instead of NULL (now the spec allows both, so not that this has to change, I just noticed the difference when testing with passing pyarrow objects to na.Schema(..))